packages/ai-openai/src/browser/openai-frontend-application-contribution.ts

```diff
@@ -38,72 +40,119 @@ export class OpenAiFrontendApplicationContribution implements FrontendApplicatio
+                    this.handleRequestSettingsChanges(event.newValue as RequestSetting[]);
                 }
             });
         });
     }
-}
 
-const openAIModelsWithDisabledStreaming = ['o1-preview'];
+    protected handleModelChanges(newModels: string[]): void {
+        const oldModels = new Set(this.prevModels);
+        const updatedModels = new Set(newModels);
 
-function createOpenAIModelDescription(modelId: string): OpenAiModelDescription {
-    return {
-        id: `openai/${modelId}`,
-        model: modelId,
-        apiKey: true,
-        enableStreaming: !openAIModelsWithDisabledStreaming.includes(modelId)
-    };
-}
+        const modelsToRemove = [...oldModels].filter(model => !updatedModels.has(model));
+        const modelsToAdd = [...updatedModels].filter(model => !oldModels.has(model));
 
-function createCustomModelDescriptionsFromPreferences(preferences: Partial<OpenAiModelDescription>[]): OpenAiModelDescription[] {
-    return preferences.reduce((acc, pref) => {
-        if (!pref.model || !pref.url || typeof pref.model !== 'string' || typeof pref.url !== 'string') {
-            return acc;
-        }
-        return [
-            ...acc,
-            {
-                id: pref.id && typeof pref.id === 'string' ? pref.id : pref.model,
-                model: pref.model,
-                url: pref.url,
-                apiKey: typeof pref.apiKey === 'string' || pref.apiKey === true ? pref.apiKey : undefined,
-                enableStreaming: pref.enableStreaming ?? true
+        this.manager.removeLanguageModels(...modelsToRemove.map(model => `openai/${model}`));
+        const requestSettings = this.preferenceService.get<RequestSetting[]>(PREFERENCE_NAME_REQUEST_SETTINGS, []);
+        this.manager.createOrUpdateLanguageModels(...modelsToAdd.map(modelId => this.createOpenAIModelDescription(modelId, requestSettings)));
+        this.prevModels = newModels;
+    }
+
+    protected handleCustomModelChanges(newCustomModels: Partial<OpenAiModelDescription>[]): void {
+        const oldModels = this.createCustomModelDescriptionsFromPreferences(this.prevCustomModels);
+        const newModels = this.createCustomModelDescriptionsFromPreferences(newCustomModels);
+
+        const modelsToRemove = oldModels.filter(model => !newModels.some(newModel => newModel.id === model.id));
+        const modelsToAddOrUpdate = newModels.filter(newModel =>
+            !oldModels.some(model =>
+                model.id === newModel.id &&
+                model.model === newModel.model &&
+                model.url === newModel.url &&
+                model.apiKey === newModel.apiKey &&
+                model.defaultRequestSettings === newModel.defaultRequestSettings &&
+                model.enableStreaming === newModel.enableStreaming));
+
+        this.manager.removeLanguageModels(...modelsToRemove.map(model => model.id));
+        this.manager.createOrUpdateLanguageModels(...modelsToAddOrUpdate);
```

Comment:
At this point of time the `requestSettings` were never mixed into the `modelsToAddOrUpdate` and will therefore be lost.

Reply:
I don't get this comment, they should be added in createCustomModelDescriptionsFromPreferences?

Reply:
The `createCustomModelDescriptionsFromPreferences` is called without the requestSettings handed over, therefore they are not retrieved and matched

Reply:
see line 126, but I will test again

Reply:
In line 126 the `requestSettings` handed over to `createCustomModelDescriptionsFromPreferences` is used. However in that method `requestSettings` is an optional parameter with a default value of an empty array. In line 79 and 80, we call `createCustomModelDescriptionsFromPreferences` without handing over the `requestSettings` parameter. Therefore `requestSettings` will be empty and all models will be updated with an `undefined` `defaultRequestSettings`

Reply:
Fixed